README.md
[![R-CMD-check](https://github.com/KWB-R/r2q/workflows/R-CMD-check/badge.svg)](https://github.com/KWB-R/r2q/actions?query=workflow%3AR-CMD-check)
[![pkgdown](https://github.com/KWB-R/r2q/workflows/pkgdown/badge.svg)](https://github.com/KWB-R/r2q/actions?query=workflow%3Apkgdown)
[![codecov](https://codecov.io/github/KWB-R/r2q/branch/main/graphs/badge.svg)](https://codecov.io/github/KWB-R/r2q)
[![Project Status](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/r2q)]()
[![R-Universe_Status_Badge](https://kwb-r.r-universe.dev/badges/r2q)](https://kwb-r.r-universe.dev/)

# r2q

The R package is used to define a tolerable pollutant input into small
surface waters via rainwater runoff. It assigns a maximal connectable urban
area to the surface water. For planning areas, different scenarios regarding
the connection of surfaces to the separate sewer system and runoff water
treatment can be calculated.

## Installation

For details on how to install KWB-R packages checkout our [installation tutorial](https://kwb-r.github.io/kwb.pkgbuild/articles/install.html).

```r
### Optionally: specify GitHub Personal Access Token (GITHUB_PAT)
### See here why this might be important for you:
### https://kwb-r.github.io/kwb.pkgbuild/articles/install.html#set-your-github_pat

# Sys.setenv(GITHUB_PAT = "mysecret_access_token")

# Install package "remotes" from CRAN
if (! require("remotes")) {
  install.packages("remotes", repos = "https://cloud.r-project.org")
}

# Install KWB package 'r2q' from GitHub
remotes::install_github("KWB-R/r2q")
```
## Example 
<details>
  <summary>Example for using the Excel input mask</summary>
  
  ```r
  path <- file.path(system.file(package = "r2q"), "extdata", "Example")
  fileName <- "Herne_Baukau.xlsx"

  # checkout the example excel file (only works on windows with excel installed)
  shell.exec(file.path(path, fileName))

  # load Example -----------------------------------------------------------
  siteData <- r2q::load_site_data(
    data.dir = path, 
    filename = fileName
  )

  c_river <- r2q::load_background_data(
    data.dir = path,
    filename = fileName, 
    default_for_na = TRUE
  )

  # load package data ---------------------------------------------------------
  c_storm <- r2q::get_stormwaterRunoff(
    runoff_effective_mix = list(
      siteData$landuse$Mix_flow_connected[c(2,1,3,4)], 
      siteData$landuse$Mix_flow_connectable[c(2,1,3,4)]),
    mix_names = c("is", "pot"))

  c_threshold <- r2q::get_thresholds(LAWA_type = siteData$LAWA_type$Value)

  # yearly rain event
  rain <- r2q::get_rain(
    area_catch = siteData$area_catch$Value, 
    river_cross_section = siteData$river_cross_section$Value,
    river_length = siteData$river_length$Value, 
    river_mean_flow = siteData$Q_mean$Value,
    x_coordinate = siteData$x_coordinate$Value,
    y_coordinate = siteData$y_coordinate$Value
  )

  # combine data ---------------------------------------------------------
  c_table <- r2q::combine_concentration_tables(
    threshold_table = c_threshold, 
    storm_table = c_storm, 
    background_table = c_river
  )

  # process -----------------------------------------------------------------
  r2q_h <- r2q::hydrology_assessment(site_data = siteData, q_rain = rain[2])
  c_type <- "average" # --> median or "worstcase" -> 95th Quantile

  # which substance poses a high risk?
  checked <- r2q::check_all_substances(
    c_table = c_table, 
    c_type = c_type)

  # Assessment of one substance
    r2q::immission_assessment(
      site_data = siteData, 
      c_table = c_table, 
      q_rain = rain[2], 
      t_rain = rain[1] * 60, substance = "Zink_geloest", 
      hazard_list = checked)

  # Assessment of all substances of potentially high risk
  r2q_out <- r2q::assess_all_hazards(
    hazard_list = checked, 
    site_data = siteData, 
    c_table = c_table, 
    q_rain = rain[2], t_rain = rain[1] * 60, 
    c_type = c_type)

  # plot the results
  r2q::plot_hazards(hazards = checked) # substances that might pose a high risk

  r2q::plot_connectable_urban_area(
    r2q_substance = r2q_out, 
    r2q_hydrology = r2q_h, 
    site_data = siteData, 
    x_type = "percent", 
    language = "de"
  )

  # detailed planning (excel sheet: "planning_area_details") ------------------
  planningData <- r2q::load_planning_details(
    data.dir = path, 
    filename = fileName,
    scenario_name = "planning_area_details" # excel sheet name of planning details
  )

  pl_out <- r2q::planning_area_discharge(
    planning_data = planningData, 
    q_rain = rain[2], 
    t_rain = rain[1] * 60, 
    y_rain = siteData$rain_year$Value, 
    thresholdTable = c_threshold)

  # comparison with tolerable discharge
  isPlan <- data.frame(
    "loadPlan_is" = pl_out$sum)
  isPlan$substance <- rownames(isPlan)
  df_compare <- merge(
    x = isPlan, 
    y = as.data.frame(lapply(r2q_out$general, unlist)), 
    by = "substance", 
    all.y = TRUE)
  df_compare
  ```
</details>

## Documentation

Release: [https://kwb-r.github.io/r2q](https://kwb-r.github.io/r2q)

Development: [https://kwb-r.github.io/r2q/dev](https://kwb-r.github.io/r2q/dev)

## Documentation

Release: [https://kwb-r.github.io/r2q](https://kwb-r.github.io/r2q)

Development: [https://kwb-r.github.io/r2q/dev](https://kwb-r.github.io/r2q/dev)


NEWS.md
# v0.1.1 (2023-07-14)

* Fix issues related to character encoding, probably due to different behaviour
  of `read.table()` on different R versions, see https://github.com/KWB-R/r2q/pull/16

# v0.1.0

First release

# v0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.

* see https://style.tidyverse.org/news.html for writing a good `NEWS.md`




R/assessments.R
#' Runs function [immission_assessment()] for all substances that might
#' pose a risk and returns the results in three tables
#' 
#' @param hazard_list Hazard list created by function [check_all_substances()]
#' @param site_data Site data list created by function [load_site_data()]
#' @param c_table Table with concentions in the river, in rainwater runoff and
#' threshold values
#' @param q_rain Intensity of a rain event in L/(ha*s)
#' @param t_rain Length of a rain event in s
#' @param c_type Character value specifiyng the type of concentration that
#' is used for the assessment. Either "average" for median value or "worstcase"
#' for 95th quantile concentration in rainwater runoff.
#' 
#' @return A list of three tables. 1) General information about the whole 
#' urbanised area calculated with data of the landuse types. 
#' 2) Detailad information about connectable area of the planning area taking
#' into account the status quo of the surrounding urbanised area. 3) Detailad 
#' information about the connectable area of the planning area, scaled down from
#' the overall connectable area.
#' 
#' @export
#' 
assess_all_hazards <- function(
    hazard_list, site_data, c_table, q_rain, t_rain, c_type = "average"
)
{
  pos_constrains <- names(hazard_list)[which(sapply(hazard_list, function(x){
    x[grep(pattern = "^pot", x = names(x))] == 1
  }))]
  
  data_out <- lapply(
    pos_constrains,
    immission_assessment, 
    site_data = site_data, c_table = c_table, hazard_list = hazard_list,
    q_rain = q_rain, t_rain = t_rain, c_type = c_type)
  
  names(data_out) <- pos_constrains
  
  general_output <- data.frame(t(sapply(data_out, function(x){
    list_out <- x[c(1,2,4:11,14)]
    names(list_out) <- 
      c("substance", "load_unit", "areaUrban_is_ha", "areaUrban_is_rel", 
        "loadUrban_is", "areaUrban_pot_ha", "area_pot_rel", "areaPlan_pot_ha", 
        "loadUrban_pot", "loadPlan_pot", "loadPlan_scaled")
    list_out
  })))
  
  planing_details_pot <- data.frame(t(sapply(data_out, function(x){
    v_out <- unlist(x[12:13])
    names(v_out) <- 
      paste0(c("res_city", "res_sub", "commercial", "main_road"), 
             c(rep("_ha", 4), rep("_rel", 4)))
    v_out
  })))
  planing_details_pot <- cbind(
    data.frame("substance" = pos_constrains, 
               planing_details_pot))
  
  planing_details_scaled <- data.frame(
    t(sapply(data_out, function(x){
      v_out <- unlist(x[15:16])
      names(v_out) <- 
        paste0(c("res_city", "res_sub", "commercial", "main_road"), 
               c(rep("_ha", 4), rep("_rel", 4)))
      v_out
    })))
  
  planing_details_scaled <- cbind(
    data.frame("substance" = pos_constrains, 
               planing_details_scaled))
  
  list("general" = general_output, 
       "plannig_pot" = planing_details_pot, 
       "planning_scaled" = planing_details_scaled)
}

#' Automated immission assessment
#' 
#' This functions works with a site data list created by function 
#' [load_site_data()], with a concentration table (see details), and with 
#' a hazard list created by function [check_all_substances()]
#' 
#' @param site_data Site data list created by function [load_site_data()]
#' @param c_table Table with concentions in the river, in rainwater runoff and
#' threshold values
#' @param q_rain Intensity of a rain event in L/(ha*s)
#' @param t_rain Length of a rain event in s
#' @param substance Substance name (as definied in c_table)
#' @param hazard_list Hazard list created by function [check_all_substances()]
#' @param c_type Character value specifiyng the type of concentration that
#' is used for the assessment. Either "average" for median value or "worstcase"
#' for 95th quantile concentration in rainwater runoff.
#' 
#' @details
#' Something about the c_table
#' 
#' @return 
#' List with all R2Q Immission assessment output value for the substance
#' 
#' @export
#'
immission_assessment <- function(
    site_data, c_table, q_rain, t_rain, substance, hazard_list, c_type = "average"
)
{
  # pre definitions
  s_row <- which(c_table$substance == substance)
  x <- hazard_list[[substance]]
  suf <- if(c_type == "average"){
    "_med"
  } else if(c_type == "worstcase"){
    "_q95"
  }
  hazard_status <- hazard_list[[substance]][paste0("pot", suf)]
  th_type <- c_table$threshold_type[which(c_table$substance == substance)]
  
  if(th_type == "acute"){
    mu <- massUnit_tranformation(  # mass unit for pollutant loads
      original_unit = unlist(strsplit(c_table$unit[s_row], split = "/"))[1], 
      change = 1)
    
    area_is <- site_data$area_urban_connected$Value * 100
    load_is <- Input_event(
      area_runoff = area_is, 
      Ci_storm = c_table[s_row, paste0("is", suf)], 
      coeff_runoff = site_data$f_D_is$Value, 
      q_rain = q_rain, 
      t_rain = t_rain
    )
    
    if(hazard_status == 1L){
      h_description <- paste0(
        substance, " might be a constraint for the implementation of seperate",
        " sewers in the planning area.")
      
      area_pot_mix <- maxArea_event(
        Q_river = site_data[["Q_event"]]$Value, 
        Ci_river = c_table$c_river[s_row], 
        Ci_threshold = c_table$threshold[s_row], 
        Ci_storm = c_table[s_row, paste0("pot", suf)], 
        coeff_runoff = site_data$f_D_pot$Value,
        q_rain = q_rain, 
        t_rain = t_rain, 
        river_length = site_data[["river_length"]]$Value,
        river_cross_section = site_data[["cross_section_event"]]$Value)
      
      area_pot_mix_rel <- area_pot_mix / 
        (site_data$area_urban_connectable$Value * 100)
      area_pot_mix_plan <- area_pot_mix_rel * site_data$area_plan$Value * 100
      
      load_pot_mix <- Input_event(
        area_runoff = area_pot_mix, 
        Ci_storm = c_table[s_row, paste0("pot", suf)], 
        coeff_runoff = site_data$f_D_pot$Value, 
        q_rain = q_rain, 
        t_rain = t_rain
      )
      
      load_available_is <- load_pot_mix - load_is
      if(load_available_is > 0){
        area_available_is <- area_from_load(
          load_runoff = load_available_is, 
          Ci_storm =  unlist(
            c_table[s_row, paste0(rownames(site_data$landuse)[1:4], suf)]),
          coeff_runoff = site_data$landuse$fD[1:4],
          q_rain = q_rain, 
          t_rain = t_rain)
        
        area_available_is_rel <- 
          area_available_is / (site_data$area_plan$Value * 100)
      } else {
        load_available_is <- 0 
        area_available_is <- rep(0, 4)
        area_available_is_rel <- rep(0, 4)
      }
      
      load_available_fair <- 
        load_pot_mix * site_data$area_plan$Value / site_data$area_urban$Value
      if(load_available_fair > 0){
        area_available_fair <- area_from_load(
          load_runoff = load_available_fair, 
          Ci_storm =  unlist(
            c_table[s_row, paste0(rownames(site_data$landuse)[1:4], suf)]),
          coeff_runoff = site_data$landuse$fD[1:4],
          q_rain = q_rain, 
          t_rain = t_rain)
        
        area_available_fair_rel <- 
          area_available_fair / (site_data$area_plan$Value * 100)
      } else {
        load_available_fair <- 0 
        area_available_fair <- rep(0, 4)
        area_available_fair_rel <- rep(0, 4)
      }
      
    } else {
      h_description <- if(hazard_status == Inf){
        paste0(
          substance, " does not pose any high risk for the surface water"
        )
      } else if(hazard_status == -Inf){
        paste0(
          "Prevent ", substance, " from entering the surface water. ",
          "Upstream concentration is already too high and stormwater " ,
          "runoff increases the risk."
        )
      }
      area_pot_mix <- area_pot_mix_rel <- area_pot_mix_plan <- load_pot_mix <- 
        load_available_is <- area_available_is <- area_available_is_rel <- 
        load_available_fair <- area_available_fair <- 
        area_available_fair_rel <- NA
    } 
    
  } else if(th_type == "annual"){
    # mass unit for pollutant loads
    mu <- massUnit_tranformation(
      original_unit = unlist(strsplit(c_table$unit[s_row], split = "/"))[1], 
      change = 2)
    area_is <- site_data$area_urban_connected$Value * 100
    
    load_is <- Input_event(
      area_runoff = area_is, 
      Ci_storm = c_table[s_row, paste0("is", suf)], 
      coeff_runoff = site_data$f_D_is$Value, 
      q_rain = site_data$rain_year$Value * 10000 / (60 * 60 * 24 * 365), 
      t_rain = 60 * 60 * 24 * 365
    ) / 1000
    if(hazard_status == 1L){
      h_description <- paste0(
        substance, " might be a constraint for the implementation of seperate",
        " sewers in the planning area.")
      
      load_pot_mix <- maxInput_year(
        Q_river = site_data[["Q_mean"]]$Value, 
        Ci_river = c_table$c_river[s_row],
        Ci_threshold = c_table$threshold[s_row], 
        Ci_storm = c_table[s_row, paste0("pot", suf)]
      ) 
      
      area_pot_mix <- maxArea_year(
        load_max = load_pot_mix, 
        Ci_threshold = c_table$threshold[s_row], 
        Ci_storm = c_table[s_row, paste0("pot", suf)], 
        coeff_runoff = site_data$f_D_pot$Value, 
        Q_rain = site_data$rain_year$Value)
      
      area_pot_mix_rel <- area_pot_mix / 
        (site_data$area_urban_connectable$Value * 100)
      area_pot_mix_plan <- area_pot_mix_rel * site_data$area_plan$Value * 100
      
      load_available_is <- load_pot_mix - load_is
      if(load_available_is > 0){
        area_available_is <- area_from_load(
          load_runoff = load_available_is * 1000, 
          Ci_storm =  unlist(
            c_table[s_row, paste0(rownames(site_data$landuse)[1:4], suf)]),
          coeff_runoff = site_data$landuse$fD[1:4],
          q_rain = site_data$rain_year$Value * 10000 / (60 * 60 * 24 * 365), 
          t_rain = 60 * 60 * 24 * 365)
        
        area_available_is_rel <- 
          area_available_is / (site_data$area_plan$Value * 100)
      } else {
        load_available_is <- 0 
        area_available_is <- rep(0, 4)
        area_available_is_rel <- rep(0, 4)
      }
      
      load_available_fair <- 
        load_pot_mix * site_data$area_plan$Value / site_data$area_urban$Value
      if(load_available_fair > 0){
        area_available_fair <- area_from_load(
          load_runoff = load_available_fair * 1000, 
          Ci_storm = unlist(
            c_table[s_row, paste0(rownames(site_data$landuse)[1:4], suf)]),
          coeff_runoff = site_data$landuse$fD[1:4],
          q_rain = site_data$rain_year$Value * 10000 / (60 * 60 * 24 * 365), 
          t_rain = 60 * 60 * 24 * 365)
        
        area_available_fair_rel <- 
          area_available_fair / (site_data$area_plan$Value * 100)
      } else {
        load_available_fair <- 0 
        area_available_fair <- rep(0, 4)
        area_available_fair_rel <- rep(0, 4)
      } 
      
    } else {
      h_description <- if(hazard_status == Inf){
        paste0(
          substance, " does not pose any high risk for the surface water"
        )
      } else if(hazard_status == -Inf){
        paste0(
          "Prevent ", substance, " from entering the surface water. ",
          "Upstream concentration is already too high and stormwater " ,
          "runoff increases the risk."
        )
      }
      area_pot_mix <- area_pot_mix_rel <- area_pot_mix_plan <- load_pot_mix <- 
        load_available_is <- area_available_is <- area_available_is_rel <- 
        load_available_fair <- area_available_fair <- 
        area_available_fair_rel <- NA
    } 
  }
  data_out <- list(
    substance,
    paste0(mu, ifelse(th_type == "acute", yes = "/Event", no = "/Year")),
    h_description,
    area_is,
    round(area_is / site_data$area_urban_connectable$Value, 0),
    signif(load_is, 4),
    round(area_pot_mix, 2),
    round(area_pot_mix_rel * 100, 0),
    round(area_pot_mix_plan, 2),
    signif(load_pot_mix,4),
    signif(load_available_is, 4),
    round(area_available_is, 2),
    round(area_available_is_rel * 100, 0),
    signif(load_available_fair, 4),
    round(area_available_fair, 2),
    round(area_available_fair_rel * 100, 0))
  
  names(data_out) <- c(
    "Substance",
    "Unit for loads",
    "Description", 
    "Currently connected area ha" ,
    "Currently connected area %" ,
    paste("Currently discharged pollutant", mu),
    "Connectable area mix ha (urban area)",
    "Connectable area mix %",
    "Connectable area mix ha (planning area)",
    paste("Tolerable pollutant load", mu, "(urban area)"),
    paste("Tolerable pollutant load", mu, "(planning area)"),
    "Connectable landuse type ha (planning area)",
    "Connectable landuse type % (planning area)",
    paste("Scaled tolerable pollutant load", mu, "(planning area)"),
    "Scaled connectable landuse type ha (planning area)",
    "Scaled connectable landuse type % (planning area)")
  
  data_out
}

#' Maximal connectable impervious area based on hydrologic conditions
#' 
#' @param site_data The site specific data loaded with function "loda_site_data"
#' @param q_rain characteristic rainfall in L/(s*ha)
#' 
#' @return 
#' the combined max_area table is extend by a row with the result of the
#' hydolic assessment.
#' 
#' @details 
#' If the planning area is identical with the urban area lines 3, 4 and 7 are
#' not valid. Line 8 is the required throttel for both, the urban area and the
#' planning area.
#' The reason for this is: for the planning area "no_runoff" landuse considered
#' for discharge calculation (-> possible runoff areas in the future), while 
#' this is not the case for the urban area (-> Status quo assessment). If the
#' planning area is no "real" planning area, that would not make any sense.
#' 
#' @export
#' 
hydrology_assessment <- function(
  site_data, q_rain
){
  
  # tolerable discharge
  Q_tolerable <- calculate_tolerable_discharge(
    area_catch = site_data[["area_catch"]]$Value, 
    area_urban = site_data[["area_urban_connectable"]]$Value, # Wording DWA: "angeschlossene befestigte Fläche des geschlossenen Siedlungsgebiets"
    area_plan = site_data[["area_plan"]]$Value, 
    area_urban_upstream = site_data[["area_urban_upstream"]]$Value,
    slope_catch = site_data[["slope_catch"]]$Value, 
    Hq1pnat_catch = site_data[["Hq1pnat_catch"]]$Value, 
    Hq2pnat_catch = site_data[["Hq2pnat_catch"]]$Value,
    verbose = T
  )
  
  # status quo discharge
  Q_tolerable$urban_is <- site_data[["area_urban_connected"]]$Value * 100 *
    q_rain * 
    site_data[["f_D_is"]]$Value
  
  Q_tolerable$planning_pot <- 
    Q_tolerable$urban - Q_tolerable$urban_is
  
  if(Q_tolerable$planning_pot < 0){
    Q_tolerable$planning_pot <- 0
  }
  
  # connectable area in ha
  area_urban <- get_allowed_area(
    f_D = site_data[["f_D_pot"]]$Value, 
    Q_tol = Q_tolerable$urban, 
    q_rain = q_rain)
  
  # connectable in %
  area_percent <- area_urban / site_data[["area_urban_connectable"]]$Value 
  
  # Discharge from Planning Area if completely connected in L/s
  max_discharge_planning <- site_data[["area_plan"]]$Value * 100 * 
    q_rain * 
    site_data$landuse$fD[-5]
  names(max_discharge_planning) <- rownames(site_data$landuse)[-5]
  
  # Potentially connectable in planning area
  discharge_planning_pot_rel <- 
    round((Q_tolerable$urban - Q_tolerable$urban_is) / 
            max_discharge_planning * 100, 1)
  discharge_planning_pot_rel[discharge_planning_pot_rel < 0] <- 0
  
  # Potentially connectable in planning area (scaled)
  discharge_planning_scaled_rel <- 
    round((Q_tolerable$planning_scaled) / 
            max_discharge_planning * 100, 1)
  
  #  Required throttel discharge in L/(s*ha) if complete planning area was connected
  throttel_scaled <- Q_tolerable$planning_scaled / 
    (site_data[["area_plan"]]$Value * 100)
  throttel_statusQuo <- Q_tolerable$planning_pot / 
    (site_data[["area_plan"]]$Value * 100)
  
  list("discharge_parameters" = data.frame(
    "Parameter" = 
      c("Tolerable discharge of the urban area",
        "Status quo discharge of the urban area",
        "Tolerable discharge of planning area (scaled)",
        "Tolerable discharge of planning area (status quo)",
        "Connetcable area in urban area (potential mix)",
        "Connectable area in percent (potential mix)",
        "Required throttel in planning area (scaled)",
        "Required throttel in planning area (status quo)"),
    "Unit" = c(rep("L/s", 4), "ha", "%", rep("L/(s*ha)", 2)),
    "Value" = c(signif(Q_tolerable$urban,3), 
                signif(Q_tolerable$urban_is, 3),
                signif(Q_tolerable$planning_scaled,3), 
                signif(Q_tolerable$planning_pot,3),
                signif(area_urban,2), 
                round(area_percent,2),
                signif(throttel_scaled,2),
                signif(throttel_statusQuo,2)),
    "Comment" = c(rep("", 6),
                  rep(paste("If the whole planning area was connected to", 
                            "the separate sewer system"),2))), 
    "planning_pot_percent" = discharge_planning_pot_rel,
    "planning_scaled_percent" = discharge_planning_scaled_rel)
}

#' Check if substances pose a risk to the surface water
#' 
#' @param c_table Table of concentration (rainwater, river, threshold value)
#' @param c_type Character value specifiyng the type of concentration that
#' is used for the assessment. Either "average" for median value or "worstcase"
#' for 95th quantile concentration in rainwater runoff.
#' 
#' @return A list with all substances defined in c_table and assigning either 1
#' or TRUE, if the substance might pose a risk, "Inf" if the substance does not
#' constrain the seperate sewer connection, and "-Inf" if ther substance should
#' not be discharged into the surface water at all, because concentration is 
#' already too high.
#' 
#' @export
#' 
check_all_substances <- function(
    c_table, c_type = "average"
)
{
  lTypes <- grep(pattern = "_med$", x = colnames(c_table), value = TRUE) 
  lTypes <- sapply(
    lTypes, 
    substr_reverse, rev_start = 1, rev_stop = 4, keep = FALSE, 
    USE.NAMES = FALSE
  )
  
  suffix <- if(c_type == "average"){
    "_med"
  } else if(c_type == "worstcase"){
    "_q95"
  }
  
  lTypes <- paste0(lTypes, suffix)
  substances <- c_table$substance
  
  check_list <- lapply(substances, function(s){
    sapply(lTypes, function(l){
      as.numeric(check_pollutant_impact(
        Ci_river = c_table$c_river[c_table$substance == s], 
        Ci_threshold = c_table$threshold[c_table$substance == s], 
        Ci_storm = c_table[c_table$substance == s, l]))
    })
  })
  names(check_list) <- substances
  check_list
}
